### PR TITLE
feat(eslint-plugin): [lines-around-comment] add extension rule

### LINKED ISSUE #1150: [lines-around-comment] Does not respect interfaces
I am not sure where to open this since the rule is not from `typescript-eslint` but since it only happens with TypeScript interfaces I am looking for direction.

**Repro**

```JSON
{
  "rules": {
    "lines-around-comment": [
      "error",
      {
        "beforeBlockComment": true,
        "afterBlockComment": false,
        "beforeLineComment": true,
        "afterLineComment": false,
        "allowBlockStart": true,
        "allowBlockEnd": false,
        "allowObjectStart": true,
        "allowObjectEnd": false,
        "allowArrayStart": true,
        "allowArrayEnd": false,
        "allowClassStart": true,
        "allowClassEnd": false,
      },
    ],
  }
}
```

```TS
// your repro code case

export interface IStartCommandOptions {
  /**
   * This will throw an error using
   */
  dangerouslyDisableHostCheck: boolean;
}
```

**Expected Result**
Shouldn't fail using interfaces

**Actual Result**

```
error  Expected line before comment  lines-around-comment
```

**Additional Info**

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `2.5.0` |
| `@typescript-eslint/parser`        | `2.5.0` |
| `TypeScript`                       | `3.6.4` |
| `ESLint`                           | `6.6.0` |
| `node`                             | `10.16.1` |
| `npm`                              | `6.12.0` |

### PR DESCRIPTION
**Pull request by chbdetta**
## PR Checklist

-   [x] Addresses an existing open issue: fixes #1150
-   [x] That issue was marked as [accepting prs](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aopen+is%3Aissue+label%3A%22accepting+prs%22)
-   [x] Steps in [CONTRIBUTING.md](https://github.com/typescript-eslint/typescript-eslint/blob/main/CONTRIBUTING.md) were taken

## Overview

port the `eslint/lines-around-comment` rule and support interface and type literals

### Options
in addition to the base rule options, this adds:
```typescript
allowInterfaceStart?: boolean;
allowInterfaceEnd?: boolean;
allowTypeStart?: boolean;
allowTypeEnd?: boolean;
allowEnumStart?: boolean;
allowEnumEnd?: boolean;
allowModuleStart?: boolean;
allowModuleEnd?: boolean;
```

### CODE REVIEW

**Comment by pkuczynski** [comment]
@JoshuaKGoldberg and @bradzacher any chance to merge this in?

**Comment by bradzacher** [comment]
@pkuczynski dont ping maintainers directly. It's poor etiquette to try and bump priorities by doing this. It doesn't change our priorities, and just spams people anyone watching this PR as well as the tagged.

We have a [documented review process](https://github.com/typescript-eslint/typescript-eslint/blob/main/CONTRIBUTING.md#addressing-feedback-and-beyond) and this PR is in the queue.

We are volunteers, and have little time to allocate to everything.
We'll get to it when we have the time to allocate to review a **_4,000_** line PR.

**Comment by pkuczynski** [comment]
Dear @bradzacher apologies for my impatience. I know how much work it is to manage popular open source package, as I am maintainer myself. I just noticed quite some activity in other PRs beings commented and merged, and this felt a bit abandoned, so indeed I wanted to bring a little bit of your attention...

File: packages/eslint-plugin/src/rules/lines-around-comment.ts
Comment (JoshuaKGoldberg) [inline packages/eslint-plugin/src/rules/lines-around-comment.ts:137]: Yeah, this kind of thing is unfortunate. Filed #5439 to see if we can clean it up.

(not blocking for this PR, the `!` should be linted against once it's unnecessary)

File: packages/eslint-plugin/src/rules/lines-around-comment.ts
Comment (JoshuaKGoldberg) [inline packages/eslint-plugin/src/rules/lines-around-comment.ts:380]: Nit:

```suggestion
      { before, after }: { before?: boolean; after?: boolean },
```

That way the `let after ..., before ...;` won't be necessary later.

File: packages/eslint-plugin/src/rules/lines-around-comment.ts
Comment (JoshuaKGoldberg) [inline packages/eslint-plugin/src/rules/lines-around-comment.ts:333]: The base rule already covers arrays, classes, and basically everything else that's not an interface or type. So you should be able to do what other rule extensions do and call the base rule to reuse its logic:

```ts
    const rules = baseRule.create(context, [options]);

    return {
      Program(): void {
        rules.Program();

        comments.forEach(...);
```

I prototyped this locally and it looks to get pretty much all the existing test cases. Is there any reason why this strategy wouldn't work?

It'd be a real shame to copy & paste so much core ESLint rule logic. We'd then have to maintain it all over time.

File: packages/eslint-plugin/src/rules/lines-around-comment.ts
Comment (JoshuaKGoldberg) [inline packages/eslint-plugin/src/rules/lines-around-comment.ts:174]: What case does this cover? Either it can be hit & should be tested, or can't be hit and this block can be removed.

File: packages/eslint-plugin/src/rules/lines-around-comment.ts
Comment (JoshuaKGoldberg) [inline packages/eslint-plugin/src/rules/lines-around-comment.ts:151]: Nit: I realize these JSDocs are mostly copypasta, but I think we* have a more ... refined JSDoc standard here 😄 

```suggestion
     * @returns Whether comments are on lines starting with or ending with code.
```

*We = me. I don't think we have a standard in the repo... but, this is a more succinct comment and plays better with intellisense.

**Review by JoshuaKGoldberg — CHANGES_REQUESTED** [review]
Ok! Thanks for sending this in and for all the thorough tests ported over from ESLint. This is a really great start. ❤️ 

But: I think we can reduce the lines of coded added by a bunch by reusing the base rule (`rules.Program()`).

**Comment by prescience-data** [comment]
Woo! Amazing to see a PR on this, I was about to start one until I noticed this 🥳

File: packages/eslint-plugin/src/rules/lines-around-comment.ts
Comment (chbdetta) [reply packages/eslint-plugin/src/rules/lines-around-comment.ts]: Sorry for the late response! I finally have time to take a look at this.

Since the base rule will always report false positive errors on TS codes (which is what we're fixing here) and to reuse the `Program()` from the base rule, I create a custom `context.report` function that silences errors caused by TS-codes and 
 I feed it to the base rule.

Then I can also skip checking non-TS codes since they are already handled by the base rule.

Changes in [30825c6](https://github.com/typescript-eslint/typescript-eslint/pull/5327/commits/30825c6f8072ea09a2a63737ffdd78e844401be6)

File: packages/eslint-plugin/src/rules/lines-around-comment.ts
Comment (bradzacher) [inline packages/eslint-plugin/src/rules/lines-around-comment.ts:343]: nit

```suggestion
      const interfaceStartAllowed =
          Boolean(options.allowInterfaceStart) &&
          isCommentAtInterfaceStart(token);
      const interfaceEndAllowed =
          Boolean(options.allowInterfaceEnd) && isCommentAtInterfaceEnd(token);
      const typeStartAllowed =
          Boolean(options.allowTypeStart) && isCommentAtTypeStart(token);
      const typeEndAllowed =
          Boolean(options.allowTypeEnd) && isCommentAtTypeEnd(token);
```

**Review by bradzacher — CHANGES_REQUESTED** [review]
there are a large number of uncovered lines in the rule - could you please add some more test cases to get the coverage higher?

**Comment by viveleroi** [comment]
What's the status of this? Looks like there are still some review changes pending? This change will be very helpful for us as currently the eslint rule is broken for interfaces/types which means we have to leave spaces where we don't want them.

**Comment by JoshuaKGoldberg** [comment]
@viveleroi the status is what it looks like: we posted some suggestions and are waiting on @chbdetta to address them.

Normally I'd link to our [Contributing guide](https://github.com/typescript-eslint/typescript-eslint/blob/main/CONTRIBUTING.md#pull-requests) to indicate that we normally ask folks to not post in PRs asking for updates. But I can see why you'd post now - it's been a month with no updates. I just updated #5942 to also mention PRs.

**Comment by JoshuaKGoldberg** [comment]
👋 @chbdetta just checking in, do you still have the time to work on this? No worries if not - we can always close the PR and let the issue be open for anybody else who wants to send a PR.

**Comment by chbdetta** [comment]
@JoshuaKGoldberg Thanks for bumping this! I was able to look at this again. I ran into some trouble getting the coverage report locally after merging the latest main, do you know how to resolve this?
```
Running coverage on untested files...Failed to collect coverage from /workspace/typescript-eslint/packages/eslint-plugin/src/configs/strict.ts
ERROR: 
  x Export assignment cannot be used when targeting ECMAScript modules. Consider using `export default` or another module format instead.
    ,-[/workspace/typescript-eslint/packages/eslint-plugin/src/configs/strict.ts:5:1]
```

File: packages/eslint-plugin/src/rules/lines-around-comment.ts
Comment (chbdetta) [reply packages/eslint-plugin/src/rules/lines-around-comment.ts:183]: Added test cases for this

**Comment by JoshuaKGoldberg** [comment]
...huh, I could swear I've responded to that message about testing coverage failure somewhere else. But can't find it. Anyway:

* What Node version are you running with locally? I don't reproduce this issue on Node 16 or 18
* You can always check the codecov report online: https://app.codecov.io/gh/typescript-eslint/typescript-eslint/compare/5327

**Comment by chbdetta** [comment]
@JoshuaKGoldberg @bradzacher please review the PR again. I delegate most of the non-TS cases to the base rule so I only keep the TS-related test cases.

**Review by JoshuaKGoldberg — APPROVED** [review]
All right, I poked at this as much as I can and it looks good to me! Thanks for sending this in & waiting for the review @chbdetta. 🔥 

@bradzacher I'll wait another week or two for your review since you've been looking at extension rules a bunch. Otherwise I'll be good to merge in.

**Review by bradzacher — APPROVED** [review]

**Pull request merged**
